### ISSUE TITLE
Error with basic auth

### ISSUE THREAD

**Issue by barebu** [hallh/elasticsearch-primary-balancer#4]
Hello.

I am trying to execute the command.
`node balance.js balance --simulation-time 3 --auth my_auth --index my_index-2020.08.01 http://my_host:9200`

But an error occurs in response

Error: Non-200 Code
    at IncomingMessage.res.on (/tmp/elasticsearch-primary-balancer-master/src/http.js:93:27)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:978:12)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9) { error: 
   { root_cause: [ [Object] ],
     type: 'security_exception',
     reason: 'missing authentication token for REST request [/_cluster/reroute]',
     header: { 'WWW-Authenticate': 'Basic realm="security" charset="UTF-8"' } },
  status: 401 }
/tmp/elasticsearch-primary-balancer-master/balance.js:154
      throw new Error('RELOCATION REQUEST FAILED');
      ^
Error: RELOCATION REQUEST FAILED
    at submitMove (/tmp/elasticsearch-primary-balancer-master/balance.js:154:13)
    at IncomingMessage.res.on (/tmp/elasticsearch-primary-balancer-master/src/http.js:93:18)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:978:12)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)

At the same time, dry-run works successfully without errors.
Help me please

**Comment by hallh** [comment]
Seems ES is saying that the `Authorization` header is missing. I checked the code and it should have been set, so not entirely sure what is going on here 🤔 Since the dry-run works, but the reroute fails, my guess would be that the user you are authorising as doesn't have the proper permissions. 

Could you check out the branch `debug/issue_4` and run the script in "suggest" mode? It'll print out a `curl` command that you should run manually to confirm whether the issue is actually caused by the missing header, or whether the error message is misleading.

**Comment by akomlik** [comment]
I get same exact error. Tried suggest and it prints correct curl command that works.

**Comment by hallh** [comment]
Hi @akomlik, can you share your ES version and any details you can about your RBAC setup. I have some time this week to try and reproduce and fix this issue.

**Comment by akomlik** [comment]
Greetings! We run Opendistro Security plugin 1.10.1.0 with ES OSS 7.9.1
Connect using basic auth user elastic with backend admin role (super user).
dry-run works ban balance breaks

```
$ curl -s -H "Authorization: Basic $AUTH" localhost:9200/ | jq -r .version
{
  "number": "7.9.1",
  "build_flavor": "oss",
  "build_type": "deb",
  "build_hash": "083627f112ba94dffc1232e8b42b73492789ef91",
  "build_date": "2020-09-01T21:22:21.964974Z",
  "build_snapshot": false,
  "lucene_version": "8.6.2",
  "minimum_wire_compatibility_version": "6.8.0",
  "minimum_index_compatibility_version": "6.0.0-beta1"
}

$ curl -s -H "Authorization: Basic $AUTH" localhost:9200/_cat/plugins| grep security
ip-10-120-0-237 opendistro_security         1.10.1.0
ip-10-120-0-222 opendistro_security         1.10.1.0
ip-10-120-0-205 opendistro_security         1.10.1.0
ip-10-120-0-213 opendistro_security         1.10.1.0
ip-10-120-0-199 opendistro_security         1.10.1.0
ip-10-120-0-233 opendistro_security         1.10.1.0
ip-10-120-0-208 opendistro_security         1.10.1.0
ip-10-120-0-197 opendistro_security         1.10.1.0
ip-10-120-0-204 opendistro_security         1.10.1.0

$ curl -s -H "Authorization: Basic $AUTH" localhost:9200/_opendistro/_security/api/account?pretty
{
  "user_name" : "elastic",
  "is_reserved" : true,
  "is_hidden" : false,
  "is_internal_user" : true,
  "user_requested_tenant" : null,
  "backend_roles" : [
    "admin"
  ],
  "custom_attribute_names" : [ ],
  "tenants" : {
    "elastic" : true,
    "global_tenant" : true,
    "admin_tenant" : true
  },
  "roles" : [
    "all_access",
    "own_index"
  ]
}
$ node balance.js dry-run --auth $AUTH --simulation-time 100 --index labdcbig2-arp localhost:9200

[-] No threshold specified, will try to achieve a perfect balance.

[+] Using threshold: 0.167
[+] Simulation time: 100 seconds.

[+] Current cluster state:
 ip-10-120-0-197:	2	(0.083)
 ip-10-120-0-199:	4	(0.167)
 ip-10-120-0-208:	2	(0.083)
 ip-10-120-0-222:	8	(0.333)
 ip-10-120-0-233:	0	(0.000)
 ip-10-120-0-237:	8	(0.333)

 SWAPPING [PRIMARY] ip-10-120-0-222/labdcbig2-arp:4 [REPLICA] ip-10-120-0-208/labdcbig2-arp:2
^C
$ node balance.js balance --auth $AUTH --simulation-time 100 --index labdcbig2-arp localhost:9200

[-] No threshold specified, will try to achieve a perfect balance.

[+] Using threshold: 0.167
[+] Simulation time: 100 seconds.

[+] Current cluster state:
 ip-10-120-0-197:	2	(0.083)
 ip-10-120-0-199:	4	(0.167)
 ip-10-120-0-208:	2	(0.083)
 ip-10-120-0-222:	8	(0.333)
 ip-10-120-0-233:	0	(0.000)
 ip-10-120-0-237:	8	(0.333)

 SWAPPING [PRIMARY] ip-10-120-0-237/labdcbig2-arp:20 [REPLICA] ip-10-120-0-233/labdcbig2-arp:4
Error: Non-200 Code
    at IncomingMessage.res.on (/usr/local/src/elasticsearch-primary-balancer/src/http.js:93:27)
    at emitNone (events.js:111:20)
    at IncomingMessage.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9) { error:
   { root_cause: [ [Object] ],
     type: 'security_exception',
     reason: 'no permissions for [cluster:admin/reroute] and User [name=opendistro_security_anonymous, backend_roles=[opendistro_security_anonymous_backendrole], requestedTenant=null]' },
  status: 403 }
/usr/local/src/elasticsearch-primary-balancer/balance.js:154
      throw new Error('RELOCATION REQUEST FAILED');
      ^

Error: RELOCATION REQUEST FAILED
    at submitMove (/usr/local/src/elasticsearch-primary-balancer/balance.js:154:13)
    at IncomingMessage.res.on (/usr/local/src/elasticsearch-primary-balancer/src/http.js:93:18)
    at emitNone (events.js:111:20)
    at IncomingMessage.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
 
```

**Comment by hallh** [comment]
Thanks for the info, it's very helpful. I'll check it out and see what's going on.

**Comment by hallh** [comment]
Hey @akomlik I fixed the bug causing the authentication error and included it in the `debug_4` branch. I'm not familiar with opendistro setup, so I tested on a vanilla ES cluster only.

Can you test and let me know whether this solves your problem?

**Comment by akomlik** [comment]
Thanks! it works now. 

```
$ node balance.js balance --auth elastic:$bootstrap_password localhost:9200

[-] No threshold specified, will try to achieve a perfect balance.

[+] Using threshold: 0.086
[+] Simulation time: 10 seconds.

[+] Current cluster state:
 ip-10-120-0-197:	13	(0.080)
 ip-10-120-0-206:	12	(0.074)
 ip-10-120-0-208:	12	(0.074)
 ip-10-120-0-211:	15	(0.093)
 ip-10-120-0-217:	12	(0.074)
 ip-10-120-0-222:	17	(0.105)
 ip-10-120-0-232:	11	(0.068)
 ip-10-120-0-233:	13	(0.080)
 ip-10-120-0-237:	17	(0.105)
 ip-10-120-0-241:	10	(0.062)
 ip-10-120-0-248:	16	(0.099)
 ip-10-120-0-250:	14	(0.086)

 SWAPPING [PRIMARY] ip-10-120-0-237/labdcbig2-system-events-000001:0 [REPLICA] ip-10-120-0-232/filebeat-7.9.2-2020.10.19:2
 SWAPPING [PRIMARY] ip-10-120-0-222/labdcbig2-interfaces:1 [REPLICA] ip-10-120-0-206/labdcbig2-stubs:5
 SWAPPING [PRIMARY] ip-10-120-0-237/labdcbig2-stubs:4 [REPLICA] ip-10-120-0-208/labdcbig2-stubs:7
 SWAPPING [PRIMARY] ip-10-120-0-248/filebeat-7.9.2-2020.10.19:3 [REPLICA] ip-10-120-0-217/labdcbig2-maps:1
 SWAPPING [PRIMARY] ip-10-120-0-222/labdcbig2-alerts:3 [REPLICA] ip-10-120-0-197/filebeat-7.9.2-2020.10.17:0
 SWAPPING [PRIMARY] ip-10-120-0-237/labdcbig2-devices:1 [REPLICA] ip-10-120-0-241/labdcbig2-interfaces:3
 SWAPPING [PRIMARY] ip-10-120-0-211/labdcbig2-hw-components:0 [REPLICA] ip-10-120-0-208/filebeat-7.9.2-2020.10.20:1
 SWAPPING [PRIMARY] ip-10-120-0-248/labdcbig2-values-1478669:10 [REPLICA] ip-10-120-0-241/filebeat-7.9.2-2020.10.21:0
 SWAPPING [PRIMARY] ip-10-120-0-197/labdcbig2-alerts:3 [REPLICA] ip-10-120-0-241/filebeat-7.9.2-2020.10.17:5
 SWAPPING [PRIMARY] ip-10-120-0-208/labdcbig2-values-1478673:2 [REPLICA] ip-10-120-0-241/filebeat-7.9.2-2020.10.17:5
 SWAPPING [PRIMARY] ip-10-120-0-208/labdcbig2-interfaces:0 [REPLICA] ip-10-120-0-241/labdcbig2-data-display:2
[-] Waiting for relocation to complete...

[+] Balance succeeded!


```

**Comment by hallh** [comment]
Great! Merged in #6 and closing this.
